File: motia-py/playground/steps/todo/delete_todo.step.py
"""Delete Todo Step."""

from typing import Any

from motia import ApiRequest, ApiResponse, FlowContext, Stream, api

todo_stream: Stream[dict[str, Any]] = Stream("todo")

config = {
    "name": "DeleteTodo",
    "description": "Delete a todo item",
    "flows": ["todo-app"],
    "triggers": [
        api("DELETE", "/todo"),
    ],
    "emits": [],
}


async def handler(request: ApiRequest[dict[str, Any]], ctx: FlowContext[Any]) -> ApiResponse[Any]:
    """Handle delete todo request."""
    body = request.body or {}
    todo_id = body.get("todo_id")

    ctx.logger.info("Deleting todo", body)

    if not todo_id:
        ctx.logger.error("todo_id is required")
        return ApiResponse(status=400, body={"error": "todo_id is required"})

    await todo_stream.delete("inbox", todo_id)

    ctx.logger.info("Todo deleted successfully", {"todo_id": todo_id})

    return ApiResponse(status=200, body={"success": True})

File: motia-js/playground/steps/todo/delete-todo.step.ts
import type { Handlers, StepConfig } from 'motia'
import { z } from 'zod'

export const config = {
  name: 'DeleteTodo',
  description: 'Delete a todo item',
  flows: ['todo-app'],
  triggers: [
    {
      type: 'http',
      method: 'DELETE',
      path: '/todo/:todoId',
      responseSchema: {
        200: z.object({ id: z.string() }),
        404: z.object({ error: z.string() }),
      },
    },
  ],
  enqueues: [],
  virtualSubscribes: ['todo-created'],
} as const satisfies StepConfig

export const handler: Handlers<typeof config> = async (request, { logger, streams }) => {
  const { todoId } = request.pathParams || {}

  logger.info('Deleting todo', { todoId })

  const existingTodo = await streams.todo.get('inbox', todoId)

  if (!existingTodo) {
    logger.warn('Todo not found', { todoId })

    return {
      status: 404,
      body: { error: `Todo with id ${todoId} not found` },
    }
  }

  await streams.todo.delete('inbox', todoId)

  logger.info('Todo deleted successfully', { todoId })

  return {
    status: 200,
    body: { id: todoId },
  }
}
